File: Stacks/735. Asteroid Collision/735. Asteroid Collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []

        for a in asteroids:
            while stack and a < 0 and stack[-1] > 0:
                if abs(a) == abs(stack[-1]):
                    stack.pop()
                    a = 0
                elif abs(a) > abs(stack[-1]):
                    stack.pop()
                else:
                    a = 0
            
            if a != 0:
                stack.append(a)
        
        return stack

# Time: O(n)
# Space: O(n)

File: Stacks/735. Asteroid Collision/735. Asteroid Collision.js
/**
 * @param {number[]} asteroids
 * @return {number[]}
 */
var asteroidCollision = function (asteroids) {
  let stack = [];

  for (let a of asteroids) {
    while (stack && a < 0 && stack.at(-1) > 0) {
      if (Math.abs(a) === Math.abs(stack.at(-1))) {
        stack.pop();
        a = 0;
      } else if (Math.abs(a) > Math.abs(stack.at(-1))) {
        stack.pop();
      } else {
        a = 0;
      }
    }

    if (a !== 0) stack.push(a);
  }

  return stack;
};

// Time: O(n)
// Space: O(n)
